mithril-client-cli/src/commands/cardano_db/verify.rs

```diff
@@ -0,0 +1,409 @@
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+use anyhow::{anyhow, Context};
+use chrono::Utc;
+use clap::Parser;
+use mithril_client::{
+    cardano_database_client::ImmutableFileRange,
+    common::{ImmutableFileNumber, MKProof, ProtocolMessage},
+    CardanoDatabaseSnapshot, Client, MessageBuilder, MithrilCertificate, MithrilResult,
+};
+use slog::{debug, warn, Logger};
+
+use crate::{
+    commands::{cardano_db::CardanoDbCommandsBackend, client_builder, SharedArgs},
+    configuration::{ConfigError, ConfigParameters, ConfigSource},
+    utils::{
+        self, CardanoDbUtils, ExpanderUtils, IndicatifFeedbackReceiver, ProgressOutputType,
+        ProgressPrinter,
+    },
+    CommandContext,
+};
+
+use super::download::shared_steps;
+
+/// Clap command to verify a Cardano db and its associated certificate.
+#[derive(Parser, Debug, Clone)]
+pub struct CardanoDbVerifyCommand {
+    #[arg(short, long, value_enum, default_value_t)]
+    backend: CardanoDbCommandsBackend,
+
+    #[clap(flatten)]
+    shared_args: SharedArgs,
+
+    /// Digest of the Cardano db snapshot to verify  or `latest` for the latest artifact
+    ///
+    /// Use the `list` command to get that information.
+    digest: String,
+
+    /// Directory from where the immutable will be verified.
+    #[clap(long)]
+    db_dir: Option<PathBuf>,
+
+    /// Genesis verification key to check the certificate chain.
+    #[clap(long, env = "GENESIS_VERIFICATION_KEY")]
+    genesis_verification_key: Option<String>,
+}
+
+impl CardanoDbVerifyCommand {
+    /// Is JSON output enabled
+    pub fn is_json_output_enabled(&self) -> bool {
+        self.shared_args.json
+    }
+
+    /// Main command execution
+    pub async fn execute(&self, context: CommandContext) -> MithrilResult<()> {
+        match self.backend {
+            CardanoDbCommandsBackend::V1 => Err(anyhow::anyhow!(
+                "The \"verify\" subcommand is not available for the v1, use --backend v2 instead",
+            )),
+            CardanoDbCommandsBackend::V2 => {
+                // *Important*: All parameters also available in the collected config source below MUST be
+                // obtained using the `params` argument.
+                // No need for `require` as validation of existence will be done by the `NewCardanoDbDownloadCommand` itself.
```

Comment:
This can be removed, it was only relevant in the indirection of the `cardano-database-v2 download` to `cardano-database download --backend v2` command